plat/drivers/rtc/pl031.c

```diff
@@ -0,0 +1,229 @@
+#include <uk/plat/common/irq.h>
+
+static __u64 pl031_boot_seconds;
+static __u64 pl031_base_addr;
+static int pl031_irq;
+
+/* FDT compatible property for PL031 device */
+#define PL031_COMPATIBLE	"arm,pl031"
+
+/* offsets of PL031 registers */
+#define RTC_DR		0	/* Data read */
+#define RTC_MR		0x4	/* Match register */
+#define RTC_LR		0x8	/* Load register */
+#define RTC_CR		0xc	/* Control register */
+#define RTC_IMSC	0x10	/* Interrupt mask and set register */
+#define RTC_RIS		0x14	/* Raw interrupt status */
+#define RTC_MIS		0x18	/* Masked interrupt status */
+#define RTC_ICR		0x1c	/* Interrupt clear register */
+
+#define PL031_REG(r)	(void *)(pl031_base_addr + (r))
+
+static __u32 pl031_read_raw(void)
+{
+	return ioreg_read32(PL031_REG(RTC_DR));
+}
+
+void pl031_read_time(struct rtc_time *rt)
+{
+	__u32 raw;
+
+	raw = pl031_read_raw();
+	rtc_raw_to_tm(raw, rt);
+}
+
+static __u32 pl031_boot_tick_get(void)
+{
+	return pl031_read_raw();
+}
+
+static void pl031_write_raw(__u32 val)
+{
+	ioreg_write32(PL031_REG(RTC_LR), val);
+}
+
+void pl031_write_time(struct rtc_time *rt)
+{
+	__u32 raw;
+
+	raw = rtc_tm_to_raw(rt);
+	pl031_write_raw(raw);
+}
+
+/*
+ * Set alarm interrupt by writing to RTC match register (MR).
+ */
+
+static void pl031_write_alarm_raw(__u32 alarm)
+{
+	ioreg_write32(PL031_REG(RTC_MR), alarm);
+}
+
+void pl031_write_alarm(struct rtc_time *rt)
+{
+	__u32 raw;
+
+	raw = rtc_tm_to_raw(rt);
+	pl031_write_alarm_raw(raw);
+}
+
+static __u32 pl031_read_alarm_raw(void)
+{
+	return ioreg_read32(PL031_REG(RTC_MR));
+}
+
+void pl031_read_alarm(struct rtc_time *rt)
+{
+	rtc_raw_to_tm(pl031_read_alarm_raw(), rt);
+}
+
+static void pl031_enable(unsigned int enable)
+{
+	ioreg_write32(PL031_REG(RTC_CR), enable);
+}
+
+/*
+ * Return RTC status.
+ *   1 denotes enable
+ *   0 denotes disable
+ *   -1 denotes wrong device
+ */
+int pl031_get_status(void)
+{
+	int val;
+
+	val = ioreg_read32(PL031_REG(RTC_CR));
+	val &= PL031_ENABLE;
+	return val;
+}
+
+/*
+ * Enable alarm interrupt.
+ *   enable = 1: enable alarm
+ *   enable = 0: disable alarm
+ */
+void pl031_enable_intr(unsigned int enable)
+{
+	ioreg_write32(PL031_REG(RTC_IMSC), enable);
+}
+
+static __u32 pl031_get_raw_intr_state(void)
+{
+	return ioreg_read32(PL031_REG(RTC_RIS));
+}
+
+void pl031_clear_intr(void)
+{
+	while (pl031_get_raw_intr_state())
+		ioreg_write32(PL031_REG(RTC_ICR), 1);
+}
+
+void pl031_register_alarm_handler(int (*handler)(void *))
+{
+	int rc;
+
+	rc = ukplat_irq_register(pl031_irq, handler, NULL);
+	if (rc < 0) {
+		uk_pr_err("Failed to register RTC interrupt handler\n");
+		return;
+	}
+}
+
+void pl031_init_rtc(void *dtb)
+{
+	__u64 size;
+	__u32 irq_type, hwirq, trigger_type;
+	int fdt_rtc, ret, rc;
+
+	uk_pr_info("Probing RTC...\n");
+
+	/*
+	 * Search for RTC device by compatible property name.
+	 */
+	fdt_rtc = fdt_node_offset_by_compatible(dtb, -1, PL031_COMPATIBLE);
+	if (fdt_rtc < 0) {
+		uk_pr_err("Could not find RTC device!, fdt_rtc is %d\n",
```

Comment:
No need for exclamation marks in error messages (didn't catch it earlier, sorry) 🙂